File: rust/2015/day21b/src/main.rs
use std::cmp;
use std::env;
use std::fs;
use std::io;
use std::process;

#[derive(Clone, Eq, PartialEq)]
enum ItemType {
    Armor,
    Weapon,
    Ring,
}

#[derive(Clone, Eq, PartialEq)]
struct Item {
    name: String,
    kind: ItemType,
    cost: i32,
    damage: i32,
    armor: i32,
}

#[derive(Clone, Eq, PartialEq)]
struct Player {
    hit_points: i32,
    damage: i32,
    armor: i32,
}

#[derive(Clone, Eq, PartialEq)]
enum Outcome {
    Win,
    Die,
}

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn init_store() -> Vec<Item> {
    vec![
        Item {
            name: String::from("Dagger"),
            kind: ItemType::Weapon,
            cost: 8,
            damage: 4,
            armor: 0,
        },
        Item {
            name: String::from("Shortsword"),
            kind: ItemType::Weapon,
            cost: 10,
            damage: 5,
            armor: 0,
        },
        Item {
            name: String::from("Warhammer"),
            kind: ItemType::Weapon,
            cost: 25,
            damage: 6,
            armor: 0,
        },
        Item {
            name: String::from("Longsword"),
            kind: ItemType::Weapon,
            cost: 40,
            damage: 7,
            armor: 0,
        },
        Item {
            name: String::from("Greataxe"),
            kind: ItemType::Weapon,
            cost: 74,
            damage: 8,
            armor: 0,
        },
        Item {
            name: String::from("Leather"),
            kind: ItemType::Armor,
            cost: 13,
            damage: 0,
            armor: 1,
        },
        Item {
            name: String::from("Chainmail"),
            kind: ItemType::Armor,
            cost: 31,
            damage: 0,
            armor: 2,
        },
        Item {
            name: String::from("Splintmail"),
            kind: ItemType::Armor,
            cost: 53,
            damage: 0,
            armor: 3,
        },
        Item {
            name: String::from("Bandedmail"),
            kind: ItemType::Armor,
            cost: 75,
            damage: 0,
            armor: 4,
        },
        Item {
            name: String::from("Platemail"),
            kind: ItemType::Armor,
            cost: 102,
            damage: 0,
            armor: 5,
        },
        Item {
            name: String::from("Damage +1"),
            kind: ItemType::Ring,
            cost: 25,
            damage: 1,
            armor: 0,
        },
        Item {
            name: String::from("Damage +2"),
            kind: ItemType::Ring,
            cost: 50,
            damage: 2,
            armor: 0,
        },
        Item {
            name: String::from("Damage +3"),
            kind: ItemType::Ring,
            cost: 100,
            damage: 3,
            armor: 0,
        },
        Item {
            name: String::from("Defense +1"),
            kind: ItemType::Ring,
            cost: 20,
            damage: 0,
            armor: 1,
        },
        Item {
            name: String::from("Defense +2"),
            kind: ItemType::Ring,
            cost: 40,
            damage: 0,
            armor: 2,
        },
        Item {
            name: String::from("Defense +3"),
            kind: ItemType::Ring,
            cost: 80,
            damage: 0,
            armor: 3,
        },
    ]
}

fn parse_file(filename: &str) -> io::Result<Player> {
    let content = fs::read_to_string(filename)?;
    let mut line_it = content.lines();
    if let Some(first_line) = line_it.next()
        && let Some((_, hp_str)) = first_line.split_once(": ")
        && let Some(second_line) = line_it.next()
        && let Some((_, damage_str)) = second_line.split_once(": ")
        && let Some(third_line) = line_it.next()
        && let Some((_, armor_str)) = third_line.split_once(": ")
    {
        let hit_points = hp_str.parse::<i32>().unwrap();
        let damage = damage_str.parse::<i32>().unwrap();
        let armor = armor_str.parse::<i32>().unwrap();
        return Ok(Player {
            hit_points,
            damage,
            armor,
        });
    }
    panic!("malformed input");
}

fn weapons(store: &[Item]) -> Vec<Item> {
    store
        .iter()
        .filter(|x| x.kind == ItemType::Weapon)
        .cloned()
        .collect()
}

fn armor(store: &[Item]) -> Vec<Item> {
    store
        .iter()
        .filter(|x| x.kind == ItemType::Armor)
        .cloned()
        .collect()
}

fn rings(store: &[Item]) -> Vec<Item> {
    store
        .iter()
        .filter(|x| x.kind == ItemType::Ring)
        .cloned()
        .collect()
}

fn compute_player(items: &[Item]) -> Player {
    Player {
        hit_points: 100,
        damage: items.iter().map(|x| x.damage).sum(),
        armor: items.iter().map(|x| x.armor).sum(),
    }
}

fn battle(player: &Player, boss: &Player) -> Outcome {
    let player_damage_taken_per_turn = cmp::max(1, boss.damage - player.armor);
    let boss_damage_taken_per_turn = cmp::max(1, player.damage - boss.armor);
    let turns_to_kill_player =
        (player.hit_points as f64 / player_damage_taken_per_turn as f64).ceil();
    let turns_to_kill_boss = (boss.hit_points as f64 / boss_damage_taken_per_turn as f64).ceil();
    if turns_to_kill_player < turns_to_kill_boss {
        Outcome::Die
    } else {
        Outcome::Win
    }
}

fn loadouts(store: &[Item]) -> Vec<Vec<Item>> {
    let weapons = weapons(store);
    let armors = armor(store);
    let rings = rings(store);
    let armor_combos = subsets(&armors, 0, 1);
    let ring_combos = subsets(&rings, 0, 2);
    let mut result = vec![];
    for w in &weapons {
        for a in &armor_combos {
            for r in &ring_combos {
                let mut loadout = vec![w.clone()];
                loadout.extend(a.iter().cloned());
                loadout.extend(r.iter().cloned());
                result.push(loadout);
            }
        }
    }
    result
}

fn subsets<T: Clone>(items: &[T], min: usize, max: usize) -> Vec<Vec<T>> {
    let mut result = Vec::new();
    for size in min..=max {
        combine(items, size, &mut Vec::new(), 0, &mut result);
    }
    result
}

fn combine<T: Clone>(
    items: &[T],
    size: usize,
    current: &mut Vec<T>,
    index: usize,
    result: &mut Vec<Vec<T>>,
) {
    if current.len() == size {
        result.push(current.clone());
        return;
    }
    for i in index..items.len() {
        current.push(items[i].clone());
        combine(items, size, current, i + 1, result);
        current.pop();
    }
}

fn losing_loadouts(store: &[Item], boss: &Player) -> Vec<Vec<Item>> {
    loadouts(store)
        .into_iter()
        .filter(|loadout| {
            let player = compute_player(loadout);
            battle(&player, boss) == Outcome::Die
        })
        .collect()
}

fn loadout_cost(loadout: &[Item]) -> i32 {
    loadout.iter().map(|item| item.cost).sum()
}

fn process_file(filename: &str) -> io::Result<i32> {
    let boss = parse_file(filename)?;
    let store = init_store();
    let highest_cost = losing_loadouts(&store, &boss)
        .iter()
        .map(|loadout| loadout_cost(loadout))
        .max()
        .unwrap();
    Ok(highest_cost)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day21b/src/Main.hs
module Main (main) where

import Data.Int (Int32)
import Data.List
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Parsec hiding (State)
import Text.Parsec.String
import Control.DeepSeq
import System.Clock

data ItemType = Armor | Weapon | Ring
                deriving (Eq, Show)

data Item = Item { name   :: String
                 , kind   :: ItemType
                 , cost   :: Int
                 , damage :: Int
                 , armor  :: Int
                 }

data Player = Player { playerHP     :: Int
                     , playerDamage :: Int
                     , playerArmor  :: Int
                     }

data Outcome = Win | Die
               deriving (Eq, Show)

store :: [Item]
store = [ Item { name = "Dagger",     kind = Weapon, cost = 8,   damage = 4, armor = 0 }
        , Item { name = "Shortsword", kind = Weapon, cost = 10,  damage = 5, armor = 0 }
        , Item { name = "Warhammer",  kind = Weapon, cost = 25,  damage = 6, armor = 0 }
        , Item { name = "Longsword",  kind = Weapon, cost = 40,  damage = 7, armor = 0 }
        , Item { name = "Greataxe",   kind = Weapon, cost = 74,  damage = 8, armor = 0 }
        , Item { name = "Leather",    kind = Armor,  cost = 13,  damage = 0, armor = 1 }
        , Item { name = "Chainmail",  kind = Armor,  cost = 31,  damage = 0, armor = 2 }
        , Item { name = "Splintmail", kind = Armor,  cost = 53,  damage = 0, armor = 3 }
        , Item { name = "Bandedmail", kind = Armor,  cost = 75,  damage = 0, armor = 4 }
        , Item { name = "Platemail",  kind = Armor,  cost = 102, damage = 0, armor = 5 }
        , Item { name = "Damage +1",  kind = Ring,   cost = 25,  damage = 1, armor = 0 }
        , Item { name = "Damage +2",  kind = Ring,   cost = 50,  damage = 2, armor = 0 }
        , Item { name = "Damage +3",  kind = Ring,   cost = 100, damage = 3, armor = 0 }
        , Item { name = "Defense +1", kind = Ring,   cost = 20,  damage = 0, armor = 1 }
        , Item { name = "Defense +2", kind = Ring,   cost = 40,  damage = 0, armor = 2 }
        , Item { name = "Defense +3", kind = Ring,   cost = 80,  damage = 0, armor = 3 }
        ]

weapons, armors, rings :: [Item]
weapons = filter ((== Weapon) . kind) store
armors  = filter ((== Armor) . kind) store
rings   = filter ((== Ring) . kind) store

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

computePlayer :: [Item] -> Player
computePlayer items =
    Player { playerHP = 100
           , playerDamage = sum [ damage item | item <- items ]
           , playerArmor = sum [ armor item | item <- items ]
           }

battle :: Player -> Player -> Outcome
battle player boss =
    let playerTakesDamagePerTurn = max 1 (playerDamage boss - playerArmor player)
        bossTakesDamagePerTurn = max 1 (playerDamage player - playerArmor boss)
        turnsToKillPlayer = ceiling ((fromIntegral (playerHP player) :: Float) /
                                     (fromIntegral playerTakesDamagePerTurn :: Float))
        turnsToKillBoss = ceiling ((fromIntegral (playerHP boss) :: Float) /
                                   (fromIntegral bossTakesDamagePerTurn :: Float))
    in if turnsToKillPlayer < turnsToKillBoss 
           then Die
           else Win

file :: Parser Player
file = do
       hp <- string "Hit Points: " *> many1 digit <* newline
       d <- string "Damage: " *> many1 digit <* newline
       a <- string "Armor: " *> many1 digit <* optional newline <* eof
       return Player { playerHP = read hp, playerDamage = read d, playerArmor = read a }

process :: String -> Int32
process content =
    case parse file "" content of
        Left err -> error (show err)
        Right boss -> fromIntegral $ maximum $ map loadoutCost $ winningLoadouts boss
  where
    loadouts :: [[Item]]
    loadouts = [ w : a ++ rs | w <- weapons,
                               a <- filter ((<= 1) . length) (subsequences armors),
                               rs <- filter ((<= 2) . length) (subsequences rings) ]

    winningLoadouts :: Player -> [[Item]]
    winningLoadouts boss = [ loadout | loadout <- loadouts,
                             let player = computePlayer loadout,
                             battle player boss == Die ]

    loadoutCost :: [Item] -> Int
    loadoutCost = sum . map cost


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"
main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
